### PR TITLE
Refactor/available timestamps

### PR DESCRIPTION
**Pull request by gmaretic**
Goals (and why):

AvailableTimestamps.handOut does not need to block on AvailableTimestamps.refreshBuffer (and in particular the CAS) if there are enough timestamps in the buffer, and can simply hand out the timestamps while the bound is being increased.

Implementation Description (bullets):

handOut now uses a different lock, and only tries to acquire a lock on (this) that is used for increasing and persisting the upper bound if necessary to actually hand out the timestamps, i.e., if the buffer is too small.

Concerns (what feedback would you like?):

Concurrency logic is not trivial to reason about, but we covered it in an ADR meeting so I'm confident it is sound. Only concern is whether it needs more documentation.

Where should we start reviewing?:

AvailableTimestamps

Priority (whenever / two weeks / yesterday):

Would love to merge this in this week, but need to get the mrtse fix in first.

### CODE REVIEW

**Comment by rhero** [comment]
Let's merge in the benchmarks first #1720.

File: docs/source/release_notes/release-notes.rst [hunk 1]
```diff
@@ -40,6 +40,10 @@ develop
     *    - Type
          - Change
 
+    *    - |improved|
+         - Refactored ``AvailableTimestamps`` reducing overzealous synchronization. Giving out timestamps is no longer blocking in refreshing the timestamp bound if there enough timestamps to give out with the current bound.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/1735>`__)
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: nit: if there are enough timestamps

File: timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java [hunk 2]
```diff
@@ -32,7 +34,7 @@
     private final Logger log;
 
     private volatile Throwable previousAllocationFailure;
-    private volatile boolean encounteredMultipleRunningTimestamps = false;
+    @Contended private volatile boolean encounteredMultipleRunningTimestamps = false;
```
Comment (jeremyk-91) [inline timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java]: Nice! Did this give us (more of) a benefit when you ran the benchmarks locally?

File: timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java [hunk 3]
```diff
@@ -47,42 +46,75 @@ private static void checkValidTimestampRangeRequest(long numberToHandOut) {
...
+     * Modifies lastReturnedTimestamp and possibly upperLimit (but obtains a lock on this if it does).
+     *
+     * Possible race condition with refreshBuffer:
+     * targetTimestamp > upperLimit.get(), indicating that we need to increase the timestamp bound, but upperLimit
+     * gets increased before we obtain the lock in allocateEnoughTimestampsToHandOut. This is, however, not a problem
+     * because the check in upperLimit.increaseToAtLeast will use the fresh value of upperLimit.get() and therefore
+     * exit the method without performing an unnecessary cas.
+     *
+     * @param numberToHandOut number of timestamps
+     * @return
```
Comment (jeremyk-91) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: nit: timestamp range handed out

File: timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java [hunk 4]
```diff
@@ -47,42 +46,75 @@ private static void checkValidTimestampRangeRequest(long numberToHandOut) {
...
              * It is important to minimize contention as much as possible on this path.
              */
-            targetTimestamp = lastHandedOut() + numberToHandOut;
-            timestampRange = handOutTimestamp(targetTimestamp);
+            long lastHandedOut = lastReturnedTimestamp.get();
+            targetTimestamp = lastHandedOut + numberToHandOut;
+            timestampRange = TimestampRange.createInclusiveRange(lastHandedOut + 1, targetTimestamp);
+            if (targetTimestamp > upperLimit.get()) {
+                //this is the only situation where we actually must block and wait for CAS
+                allocateEnoughTimestampsToHandOut(targetTimestamp, ALLOCATION_BUFFER_SIZE);
```
Comment (jeremyk-91) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: Yep, makes sense - the bound _must_ be at least targetTimestamp, but if we're doing a CAS let's bump it up a good bit more.

File: timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java [hunk 5]
```diff
@@ -47,42 +46,75 @@ private static void checkValidTimestampRangeRequest(long numberToHandOut) {
...
              * Under high concurrent load, this will be a hot critical section as clients request timestamps.
              * It is important to minimize contention as much as possible on this path.
              */
-            targetTimestamp = lastHandedOut() + numberToHandOut;
-            timestampRange = handOutTimestamp(targetTimestamp);
+            long lastHandedOut = lastReturnedTimestamp.get();
+            targetTimestamp = lastHandedOut + numberToHandOut;
+            timestampRange = TimestampRange.createInclusiveRange(lastHandedOut + 1, targetTimestamp);
+            if (targetTimestamp > upperLimit.get()) {
+                //this is the only situation where we actually must block and wait for CAS
```
Comment (jeremyk-91) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: nit: normally we leave a space between the comment slashes and the comment text

File: timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java (hunk 2 above)
Comment (gmaretic) [inline timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java]: Didn't get around to trying it out yet, but will let you know :)

**Comment by jeremyk-91** [comment]
We've discussed the concurrency changes and I'm pretty happy with the state of the PR now. As a final test, I'm going to run the Jepsen tests against this branch a couple of times.

**Comment by gmaretic** [comment]
@jboreiko Can we merge this in?

File: timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java (hunk 2 above)
Comment (jboreiko) [inline timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java]: heads up - [00:00:20] /home/ubuntu/atlasdb-internal-testing/timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java:26: warning: Contended is internal proprietary API and may be removed in a future release

**Comment by jboreiko** [comment]
@davidscottcohen this is the timestamp throughput perf ticket I had mentioned.

**Comment by gmaretic** [comment]
Merging this with develop seems to be way too much pain. Opened https://github.com/palantir/atlasdb/pull/1783 to get the code in

**Comment by gmaretic** [comment]
Merged as #1783

**Pull request closed without merging**